tests/cpp/test_resize.cpp

```diff
@@ -3733,4 +3730,522 @@ TEST_F(ResizeTest, SliceScheduledLikeProducer) {
+
+  std::vector<IterDomain*> tv3_loop{
+      tv2->getRootDomain()[0]->cloneWithoutRFactor(),
+  };
+
+  IrBuilder::create<Resize>(
+      tv3->getRootDomain().at(0),
+      tv3_loop.at(0),
+      IrBuilder::create<Val>(-shape[0] / 2, DataType::Index),
+      fusion.zeroVal());
+
+  tv3->setLoopDomain(tv3_loop);
+
+  for (auto tv : {tv1, tv2, tv3}) {
+    tv->split(0, 32);
+    tv->axis(0)->parallelize(ParallelType::BIDx);
+    tv->axis(1)->parallelize(ParallelType::TIDx);
+  }
+
+  auto options = at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);
+  auto t0 = at::randn(shape, options);
+  std::vector<c10::IValue> aten_inputs({t0});
+
+  FusionExecutor fe;
+  fe.compileFusion(&fusion, aten_inputs);
+  auto cg_outputs = fe.runFusion(aten_inputs);
+
+  auto ref = at::pad(
+      t0.index({at::indexing::Slice(shape[0] / 2, shape[0])}),
+      {shape[0] / 2, 0});
+
+  EXPECT_TRUE(ref.equal(cg_outputs[0]));
+}
+
+TEST_F(ResizeTest, SliceThenConcat) {
+  Fusion fusion;
+  FusionGuard fg(&fusion);
+
+  std::vector<int64_t> shape({100});
+
+  // concrete shapes to avoid dynamic Fusion
+  auto tv0 = makeContigConcreteTensor(shape);
+  fusion.addInput(tv0);
+
+  auto tv1 = set(tv0);
+
+  // TODO: Use cat instead of the manual pad + add
+
+  // left half
+  auto tv2 =
+      slice(tv1, {{fusion.zeroVal(), IrBuilder::create<Val>(shape[0] / 2)}});
+  auto tv3 = pad(tv2, {fusion.zeroVal(), IrBuilder::create<Val>(shape[0] / 2)});
+
+  // right half
+  auto tv4 = slice(
+      tv1,
+      {{IrBuilder::create<Val>(shape[0] / 2),
+        IrBuilder::create<Val>(shape[0])}});
+  auto tv5 = pad(tv4, {IrBuilder::create<Val>(shape[0] / 2), fusion.zeroVal()});
+
+  auto tv6 = add(tv3, tv5);
+
+  fusion.addOutput(tv6);
+
+  tv2->setLoopDomain(tv2->getRootDomain());
+
+  {
+    std::vector<IterDomain*> tv3_loop{
+        tv2->getRootDomain()[0]->cloneWithoutRFactor(),
+    };
+    IrBuilder::create<Resize>(
+        tv3->getRootDomain().at(0),
+        tv3_loop.at(0),
+        fusion.zeroVal(),
+        IrBuilder::create<Val>(-shape[0] / 2, DataType::Index));
+    tv3->setLoopDomain(tv3_loop);
+  }
+
+  tv4->setLoopDomain(tv4->getRootDomain());
+
+  {
+    std::vector<IterDomain*> tv5_loop{
+        tv4->getRootDomain()[0]->cloneWithoutRFactor(),
+    };
+    IrBuilder::create<Resize>(
+        tv5->getRootDomain().at(0),
+        tv5_loop.at(0),
+        IrBuilder::create<Val>(-shape[0] / 2, DataType::Index),
+        fusion.zeroVal());
+    tv5->setLoopDomain(tv5_loop);
+  }
+
+  {
+    std::vector<IterDomain*> tv6_loop{
+        tv2->getRootDomain()[0]->cloneWithoutRFactor(),
+    };
+    auto left_half = IterDomain::resize(
+        tv6_loop[0],
+        fusion.zeroVal(),
+        IrBuilder::create<Val>(-shape[0] / 2, DataType::Index));
+    IrBuilder::create<Resize>(
+        tv6->getLogicalDomain().at(0),
+        left_half,
+        fusion.zeroVal(),
+        IrBuilder::create<Val>(shape[0] / 2, DataType::Index));
+    tv6->setLoopDomain(tv6_loop);
```

Comment:
Are you asking why this won't work?

```
std::vector<IterDomain*> tv6_loop{
        tv2->getRootDomain()[0]->cloneWithoutRFactor(),
    };
tv6->setLoopDomain(tv6_loop);
```

This won't work as this `tv6_loop` domain is just a copy of the `tv2` domain and is not connected with anything yet, i.e., it doesn't have any definition or use expr. We need to define how it is related to the existing domains of `tv6`.

To connect the domain to `tv6`, we can use either the left or right half. It won't matter. I should have mentioned that by a comment.

Reply:
Please bear with more of my naive questions.

> a copy of the tv2 domain and is not connected with anything yet, i.e., it doesn't have any definition or use expr.

I thought `tv2->getRootDomain()` is consumed by a resize to create its logical domain. Since we have
`auto tv2 = slice(tv1, {{fusion.zeroVal(), IrBuilder::create<Val>(shape[0] / 2)}});` 

> We need to define how it is related to the existing domains of tv6

since we have a chain of `slice -> pad -> binary_add`. With exact graph, if we view the IterDomain as a ValGroup, I think the connection is well established.

Of course this is just theoretical discussion ^^^.

`We need to define how it is related to the existing domains of tv6` -> Does it mean our analysis today requires the chain of relationship needs to be resolved within the IterDomains of a single TensorDomain?

Reply:
> I thought tv2->getRootDomain() is consumed by a resize to create its logical domain. Since we have
auto tv2 = slice(tv1, {{fusion.zeroVal(), IrBuilder::create<Val>(shape[0] / 2)}});

That is true, but `tv2->getRootDomain()[0]->cloneWithoutRFactor()` is not. And since it's just a plain copy, it isn't mapped with anything.

> since we have a chain of slice -> pad -> binary_add. With exact graph, if we view the IterDomain as a ValGroup, I think the connection is well established.

Again, that's true with `tv2->getRootDomain()[0]` but not with its clone.

> Does it mean our analysis today requires the chain of relationship needs to be resolved within the IterDomains of a single TensorDomain?

Yes and no, depending on what you're trying to get from the analysis. In this case, however, since the clone is not mapped with anything, there's nothing we can get from it without connecting it to other domains through replays.

Reply:
>  but tv2->getRootDomain()[0]->cloneWithoutRFactor() is not. And since it's just a plain copy

:exploding_head: Ha, I see what I'm missing now. Thanks a lot for walking me through this :bow: